src/kernel/include/thd.h

```diff
@@ -39,6 +39,19 @@ struct rcvcap_info {
 	struct thread *rcvcap_thd_notif; /* The parent rcvcap thread for notifications */
 };
 
+/*
+ * This is the data structure embedded in the cos_cpu_local_info that
+ * contains information for the thread that was either preempted or woken up
+ * and is waiting to be scheduled instead of the current thread's scheduler
+ * upon RCV syscall. This is mainly to reduce the number of context switches
+ * to schedule the thread that is deemed eligible by the scheduler. 
+ */
+struct next_thdinfo {
+	struct tcap   *tc;
+	struct thread *thd;
+	tcap_prio_t    prio;
+} per_core_nextti[NUM_CPU_COS];
```

Comment:
Ahh, OK, I see.  You're using a pointer in the `cos_cpu_local_info` to this.

Why not just include a field `struct next_thdinfo thd_next;` (notice, it isn't a pointer).